GoogleDataTransport/GDTCORLibrary/GDTCORDatabase.m

```diff
@@ -0,0 +1,316 @@
+  }
+}
+
+#pragma mark - GDTCORDatabase
+
+@implementation GDTCORDatabase {
+  /** The sqlite database. */
+  sqlite3 *_db;
+}
+
+- (nullable instancetype)initWithURL:(nullable NSURL *)dbFileURL
+                 migrationStatements:
+                     (nonnull NSDictionary<NSNumber *, NSString *> *)migrationStatements {
+  // Only 1 instance per file is allowed. Return nil if one is already open.
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    instancesToDBFiles = [[NSMutableDictionary alloc] init];
+  });
+  __block GDTCORDatabase *extantDB;
+  dispatch_barrier_sync(dispatch_get_global_queue(QOS_CLASS_DEFAULT, 0), ^{
+    extantDB = instancesToDBFiles[dbFileURL.path];
+  });
+  if (extantDB) {
+    return nil;
+  }
+
+  // Create an instance.
+  self = [super init];
+  if (self) {
+    NSString *dbPath = dbFileURL ? dbFileURL.path : @":memory:";
+    if (!GDTCORSQLOpenDB(&_db, dbPath)) {
+      return nil;
+    }
+    _dbQueue = dispatch_queue_create("com.google.GDTCORDatabase", DISPATCH_QUEUE_SERIAL);
+    _fileURL = dbFileURL;
+    _stmtCache =
+        CFDictionaryCreateMutable(CFAllocatorGetDefault(), 0, &kCFTypeDictionaryKeyCallBacks, NULL);
+    dispatch_async(_dbQueue, ^{
+      RunMigrations(self->_db, [migrationStatements copy], self -> _stmtCache);
+    });
+    dispatch_barrier_sync(dispatch_get_global_queue(QOS_CLASS_DEFAULT, 0), ^{
+      instancesToDBFiles[dbPath] = self;
+    });
+  }
+  return self;
+}
+
+- (void)dealloc {
+  [self close];
+}
+
+- (BOOL)open {
+  if (_db == NULL) {
+    return GDTCORSQLOpenDB(&_db, _fileURL ? _fileURL.path : @":memory:");
+  }
+  GDTCORLogWarning(GDTCORMCWDatabaseWarning, @"Database already open: %@", _fileURL);
+  return NO;
+}
+
+- (BOOL)close {
+  if (_dbQueue != nil && _db != NULL) {
+    dispatch_sync(_dbQueue, ^{
+      GDTCORSQLCloseDB(_db);
+      _db = NULL;
+    });
+    if (_fileURL) {
+      dispatch_barrier_sync(dispatch_get_global_queue(QOS_CLASS_DEFAULT, 0), ^{
+        [instancesToDBFiles removeObjectForKey:_fileURL.path];
+      });
+    }
+    return YES;
+  }
+  return NO;
+}
+
+- (void)setUserVersion:(int)userVersion {
+  dispatch_async(_dbQueue, ^{
+    SetUserVersion(self->_db, userVersion);
+  });
+}
+
+- (int)userVersion {
+  __block int userVersion = -1;
+  dispatch_sync(_dbQueue, ^{
+    userVersion = GetUserVersion(self->_db);
+  });
+  return userVersion;
+}
+
+- (BOOL)runNonQuery:(NSString *)sql
+           bindings:(NSDictionary<NSNumber *, NSString *> *)bindings
+          cacheStmt:(BOOL)cacheStmt {
+  __block BOOL returnStatus = NO;
+  dispatch_sync(_dbQueue, ^{
+    returnStatus = RunNonQuery(_db, bindings, sql, self->_stmtCache, cacheStmt);
+  });
+  return returnStatus;
+}
+
+- (BOOL)runQuery:(NSString *)sql
+        bindings:(NSDictionary<NSNumber *, NSString *> *)bindings
+         eachRow:(GDTCORSqliteRowResultBlock)eachRow
+       cacheStmt:(BOOL)cacheStmt {
+  __block BOOL returnStatus = NO;
+  dispatch_sync(_dbQueue, ^{
```

Comment:
As far as I understand, the class is intended to be a thread-safe but it will be a client responsibility to make sure the main thread is not blocked. It is correct?

Reply:
Yeah, though I'm not sure how this class could cause a main thread deadlock.

Reply:
Ah, no I don't mean a deadlock, I mean a long operation on the main thread. Such an operation e.g. at app start may lead to a watchdog killing the app. Just wanted to confirm that it is intended for the class clients to run the methods of the main thread.

Reply:
Yeah, if the main-thread blocks waiting for the result of a read, that's the adopter's problem, not this class's (same should be true of other db classes).